pkg/package-server-manager/controller.go

```diff
@@ -0,0 +1,192 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"sync"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/operator-framework-olm/pkg/manifests"
+	olmv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	name                = "packageserver"
+	infrastructureName  = "cluster"
+	defaultReplicaCount = int32(2)
+	defaultRolloutCount = 1
+	singleReplicaCount  = int32(1)
+)
+
+// PackageServerCSVReconciler reconciles the PackageServer CSV object
+type PackageServerCSVReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+	Lock   sync.Mutex
+
+	Name                string
+	Namespace           string
+	HighlyAvailableMode bool
+}
+
+// TODO(tflannag): Determine what kubebuilder markers are required here
+func (r *PackageServerCSVReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("csv", req.NamespacedName)
+
+	log.Info("handling current request", "request", req.String())
+	log.Info("currently topology mode", "highly available", r.getHighlyAvailableMode())
+	defer log.Info("finished request reconciliation")
+
+	// TODO(tflannag): Cleanup this structure. Maybe CreateOrUpdate would help here
+	csv := &olmv1alpha1.ClusterServiceVersion{}
+	if err := r.Client.Get(ctx, req.NamespacedName, csv); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+		if apierrors.IsNotFound(err) {
+			log.Info("csv does not exist attempting to create it")
+			csv, err = manifests.NewPackageServerCSV(r.Namespace)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			_ = ensureCSV(log, csv, r.getHighlyAvailableMode())
+			if err := r.Client.Create(ctx, csv); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	if !csv.ObjectMeta.DeletionTimestamp.IsZero() {
+		log.Info("requeuing after encountering a delete event")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	orig := client.MergeFrom(csv.DeepCopy())
+	if !ensureCSV(log, csv, r.getHighlyAvailableMode()) {
+		log.Info("no updates to the csv are necessary")
```

Comment:
nit: will these logs show up in non-debug configurations? If so, I think this might end up generating a lot of noise.

Reply:
Unfortunately, there's no notion of a debug handler attached to the log.Logger that was generated at scaffolding time by kubebuilder. I believe you can instead just specify the verbosity level, so maybe `log.V(5).Info(...)` would be more appropriate?

Reply:
I guess there's also no real reason we have to use this logging library, instead of logrus, which seems to be the favorite when poking around the various OLM-related repositories.